src/note.rs

```diff
@@ -0,0 +1,394 @@
+        )?;
+        n += buf.write(&self.nonce.to_bytes())?;
+        n += buf.write(&self.stealth_address.to_bytes())?;
+        n += buf.write(&self.pos.to_le_bytes())?;
+        n += buf.write(&self.encrypted_data[..])?;
+
+        buf.flush()?;
+        Ok(n)
+    }
+}
+
+impl Write for Note {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let mut buf = io::BufReader::new(&buf[..]);
+
+        let mut one_byte = [0u8; 1];
+        let mut one_scalar = [0u8; 32];
+        let mut one_u64 = [0u8; 8];
+        let mut one_stealth_address = [0u8; 64];
+
+        let mut n = 0;
+
+        buf.read_exact(&mut one_byte)?;
+        n += one_byte.len();
+        self.note_type = one_byte[0].try_into()?;
+
+        buf.read_exact(&mut one_scalar)?;
+        n += one_scalar.len();
+        self.value_commitment =
+            JubJubExtended::from(jubjub_decode::<JubJubAffine>(&one_scalar)?);
+
+        buf.read_exact(&mut one_scalar)?;
+        n += one_scalar.len();
+        self.nonce = jubjub_decode::<JubJubScalar>(&one_scalar)?;
+
+        buf.read_exact(&mut one_stealth_address)?;
+        n += one_stealth_address.len();
+        self.stealth_address =
+            StealthAddress::from_bytes(&one_stealth_address)?;
+
+        buf.read_exact(&mut one_u64)?;
+        n += one_u64.len();
+        self.pos = u64::from_le_bytes(one_u64);
+
+        buf.read_exact(&mut self.encrypted_data)?;
+        n += self.encrypted_data.len();
+
+        Ok(n)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+impl Note {
+    /// Creates a new phoenix output note
+    pub fn new(note_type: NoteType, psk: &PublicSpendKey, value: u64) -> Self {
+        let r = JubJubScalar::random(&mut rand::thread_rng());
+        let nonce = JubJubScalar::random(&mut rand::thread_rng());
+        let blinding_factor = JubJubScalar::random(&mut rand::thread_rng());
+
+        Self::deterministic(note_type, &r, nonce, psk, value, blinding_factor)
+    }
+
+    /// Creates a new transparent note
+    pub fn transparent(psk: &PublicSpendKey, value: u64) -> Self {
+        Self::new(NoteType::Transparent, psk, value)
+    }
+
+    /// Creates a new obfuscated note
+    pub fn obfuscated(psk: &PublicSpendKey, value: u64) -> Self {
+        Self::new(NoteType::Obfuscated, psk, value)
+    }
+
+    /// Create a new phoenix output note without inner randomness
+    pub fn deterministic(
+        note_type: NoteType,
+        r: &JubJubScalar,
+        nonce: JubJubScalar,
+        psk: &PublicSpendKey,
+        value: u64,
+        blinding_factor: JubJubScalar,
+    ) -> Self {
+        let stealth_address = psk.gen_stealth_address(r);
+
+        let value_commitment = JubJubScalar::from(value);
+        let value_commitment = JubJubExtended::from(JubJubAffine::from(
+            &(GENERATOR_EXTENDED * value_commitment)
+                + &(GENERATOR_NUMS_EXTENDED * blinding_factor),
+        ));
```

Comment:
No need to do this conversions. 
```rust
let value_commitment = (GENERATOR_EXTENDED * value_commitment)
                + (GENERATOR_NUMS_EXTENDED * blinding_factor);
```

Reply:
As discussed, it was needed because the hashing. Now we're using `to_hash_inputs()` and there is no need to such conversion here.